### DISCUSSION TITLE
How to split arrays from JSON using Infinity datasource plugin

### DISCUSSION THREAD

**Discussion by asktushar** [grafana/grafana-infinity-datasource#206] [Q&A] [answered]
I used the Infinity plugin to make a call to REST API.
The following is the JSON response(Structure) with timestamp (epoch) and values.
```

{
  "report": [
    {
      "timeWindow": {
        "entityEpochSecUTC": 1625788800,
        "entityTimeDurationSec": "SECOND",
        "entityTimeField": "kafka_ingestion_time"
      },
      "entity": {
        "entityType": "hive",
        "entityNamespace": "db",
        "entityName": "b_table",
        "entityInstance": "server-ams4",
        "dataAssetRef": "hive.db.b_table-v_1.0"
      },
      "reportItems": [
        {
          "checkName": "Check for cc info",
          "checkLevel": "Error",
          "checkStatus": "Success",
          "constraintName": "containsCreditCardNumber",
          "constraintStatus": "Success",
          "constraintMessage": "string"
        }
      ]
    }
  ]
}
```

What i wish to get from the above JSON in the Grafana Dashboard is below table (Assume below are the columns of table).

timeWindow.entityEpochSecUTC - entity.entityInstance - reportItems.checkLevel - reportItems.checkStatus

Here `reportItems` is an array so it has many rows but i want it to split and give me all the rows against `timeWindow.entityEpochSecUTC`, `entity.entityInstance`

I have tried the below approach - 
 
![Screenshot 2021-12-01 at 20 05 04](https://user-images.githubusercontent.com/6053964/144297738-adc4f414-1406-461d-bea9-13ce772f0794.png)

If you see, it is able to get the reportItems completely and also `reportItems[1].checkLevel` however if i try to get `reportItems.checkLevel` it fails to render anything. 

Another approach i tried is getting reportItems from a separate query and `timeWindow.entityEpochSecUTC`, `entity.entityInstance`
from separate and them merge them. But somehow reportItems as a root, does not get the checkLevel values. 

![Screenshot 2021-12-01 at 20 10 12](https://user-images.githubusercontent.com/6053964/144298372-44882642-e27b-4efd-856e-7efe7989fe89.png)

Looking forward to understand this behaviour and get some inspiration if something like this is already dealt with.

**Comment by yesoreyeram** [comment]
Does `report[0].reportItems` as a root will solve your issue? Assuming your report array have only one item.

**Reply by asktushar** [reply]
Hi @yesoreyeram, Thanks for your reply. `reportItems` and  `report`  can have multiple items, This is just the structure i shared in the above description.

**Comment by asktushar** [comment]
Hi @yesoreyeram, Do you think this use case is possible with the infinity-datasource Plugin ?

**Reply by yesoreyeram** [reply]
With the current state, This is not possible I think.

**Comment by yesoreyeram** [comment, marked as answer]
With the recent version 0.8.6 of infinity, this is possible with UQL type query. Key part is `mv-expand` command which expands array items into individual rows.

Example query [here](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%5Cn%7B%5Cn%20%20%5C%22report%5C%22:%20%5B%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%5C%22timeWindow%5C%22:%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22entityEpochSecUTC%5C%22:%201625788800,%5Cn%20%20%20%20%20%20%20%20%5C%22entityTimeDurationSec%5C%22:%20%5C%22SECOND%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22entityTimeField%5C%22:%20%5C%22kafka_ingestion_time%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%5C%22entity%5C%22:%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22entityType%5C%22:%20%5C%22hive%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22entityNamespace%5C%22:%20%5C%22db%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22entityName%5C%22:%20%5C%22b_table%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22entityInstance%5C%22:%20%5C%22server-ams4%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22dataAssetRef%5C%22:%20%5C%22hive.db.b_table-v_1.0%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%5C%22reportItems%5C%22:%20%5B%5Cn%20%20%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22checkName%5C%22:%20%5C%22Check%20for%20cc%20info%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22checkLevel%5C%22:%20%5C%22Error%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22checkStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22constraintName%5C%22:%20%5C%22containsCreditCardNumber%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22constraintStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22constraintMessage%5C%22:%20%5C%22string%5C%22%5Cn%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%5D%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22timeWindow%5C%22:%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityEpochSecUTC%5C%22:%201625788800,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityTimeDurationSec%5C%22:%20%5C%22SECOND%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityTimeField%5C%22:%20%5C%22kafka_ingestion_time%5C%22%5Cn%20%20%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%20%20%5C%22entity%5C%22:%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityType%5C%22:%20%5C%22hive%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityNamespace%5C%22:%20%5C%22db%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityName%5C%22:%20%5C%22b_table%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22entityInstance%5C%22:%20%5C%22server-ams5%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%5C%22dataAssetRef%5C%22:%20%5C%22hive.db.b_table-v_1.0%5C%22%5Cn%20%20%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%20%20%5C%22reportItems%5C%22:%20%5B%5Cn%20%20%20%20%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkName%5C%22:%20%5C%22Check%20for%20cd%20info%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkLevel%5C%22:%20%5C%22Error%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintName%5C%22:%20%5C%22containsCreditCardNumber%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintMessage%5C%22:%20%5C%22string%5C%22%5Cn%20%20%20%20%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkName%5C%22:%20%5C%22Check%20for%20ce%20info%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkLevel%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22checkStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintName%5C%22:%20%5C%22containsCreditCardNumber%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintStatus%5C%22:%20%5C%22Success%5C%22,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22constraintMessage%5C%22:%20%5C%22string%5C%22%5Cn%20%20%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%20%20%5D%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%5D%5Cn%7D%22,%22uql%22:%22parse-json%5Cn%23%20%7C%20scope%20%5C%22report%5C%22%5Cn%23%20%7C%20mv-expand%20%5C%22reportItem%5C%22%3D%5C%22reportItems%5C%22%5Cn%23%20%7C%20project%20%5C%22timestamp%5C%22%3Dunixtime_seconds_todatetime(%5C%22timeWindow.entityEpochSecUTC%5C%22),%20%5C%22instanceName%5C%22%3D%5C%22entity.entityInstance%5C%22,%20%5C%22checkLevel%5C%22%3D%5C%22reportItem.checkLevel%5C%22,%20%5C%22status%5C%22%3D%5C%22reportItem.checkStatus%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D).
```kusto
parse-json
| scope "report"
| mv-expand "reportItem"="reportItems"
| project "timestamp"=unixtime_seconds_todatetime("timeWindow.entityEpochSecUTC"), "instanceName"="entity.entityInstance", "checkLevel"="reportItem.checkLevel", "status"="reportItem.checkStatus"
```

<img width="1687" alt="image" src="https://user-images.githubusercontent.com/153843/164413611-b2660172-775f-4480-811b-606915af6527.png">
